ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.cpp

```diff
@@ -0,0 +1,414 @@
+    const bool is_handshake_master = this->my_chip_id < this->peer_chip_id;
+    TT_ASSERT(this->my_chip_id != this->peer_chip_id);
+    TT_ASSERT(
+        this->sender_0_num_buffers == this->sender_1_num_buffers);  //, "Implementation expects sender_0_num_buffers and
+                                                                    // sender_1_num_buffers to be the same for now");
+    log_trace(tt::LogTest, "Sender 0 num buffers: {}", this->sender_0_num_buffers);
+    log_trace(tt::LogTest, "Sender 0 channel address: {}", this->local_sender_channel_0_buffer_address);
+    log_trace(tt::LogTest, "Sender 1 num buffers: {}", this->sender_1_num_buffers);
+    log_trace(tt::LogTest, "Sender 1 channel address: {}", this->local_sender_channel_1_buffer_address);
+    log_trace(tt::LogTest, "Receiver num buffers: {}", this->receiver_num_buffers);
+    log_trace(tt::LogTest, "Receiver channel address: {}", this->local_receiver_channel_buffer_address);
+    return std::vector<uint32_t>{
+        is_handshake_master,
+        this->handshake_address,
+        this->channel_buffer_size,
+
+        this->sender_0_num_buffers,
+        this->receiver_num_buffers,
+
+        config.sender_0_channel_base_address,
+        FabricEriscDatamoverConfig::sender_channel_0_worker_connection_info_address,
+        config.sender_1_channel_base_address,
+        FabricEriscDatamoverConfig::sender_channel_1_worker_connection_info_address,
+        config.receiver_channel_base_address,
+        config.receiver_channel_base_address,
+
+        config.sender_0_channel_base_address,
+        config.sender_1_channel_base_address,
+
+        this->termination_signal_ptr};
+}
+
+std::vector<uint32_t> FabricEriscDatamoverBuilder::emit_runtime_args() const {
+    return std::vector<uint32_t>{
+        this->sender_channel_0_connection_semaphore_id,
+        this->sender_channel_1_connection_semaphore_id,
+        this->sender_channel_0_buffer_index_semaphore_id,
+        this->downstream_sender_channel_buffer_index_semaphore_id.value_or(-1),
+        this->downstream_edm_buffer_base_address != std::nullopt,
+        this->downstream_edm_buffer_base_address.value_or(0),
+        this->downstream_edm_noc_x.value_or(0),
+        this->downstream_edm_noc_y.value_or(0),
+        this->downstream_edm_semaphore_address.value_or(-1),
+        this->downstream_edm_worker_registration_address.value_or(0),
+        this->downstream_edm_worker_location_info_address.value_or(0),
+        this->receiver_channel_local_buffer_index_addr,
+        // this is the receiver channel's local sem for flow controlling with downstream fabric sender
+        this->receiver_channel_downstream_flow_control_semaphore_id.value_or(0),
+        this->sender_channel_0_flow_control_semaphore_id,
+        this->sender_channel_1_flow_control_semaphore_id
+    };
+}
+
+FabricEriscDatamoverBuilder FabricEriscDatamoverBuilder::build_builder(
```

Comment:
`Builder Builder::build_builder` is a bit too much :D

Reply:
I was thinking of adding one more but I thought 4 was the tasteful number.

Reply:
can you think of something cleaner? not a blocker, but I think would be good to streamline this.

Reply:
This isn't the intended user API (though it's used in one of the tests because the test is trying a non-standard setup (loopback mode). For all other use cases, the user will just use `EdmLineFabricOpInterface`, which wraps and hides this.

Reply:
I met you ~in the middle~ a quarter of the way there and removed a "build" ;)